src/backend/distributed/transaction/distributed_transaction_management.c

```diff
@@ -0,0 +1,483 @@
+
+	/* we may update the shmem, acquire lock exclusively */
+	LWLockAcquire(AddinShmemInitLock, LW_EXCLUSIVE);
+
+	distributedTransactionShmemData =
+		(DistributedTransactionShmemData *) ShmemInitStruct(
+			"Distributed Transaction Management",
+			DistributedTransactionManagementShmemSize(),
+			&alreadyInitialized);
+
+	if (!alreadyInitialized)
+	{
+#if (PG_VERSION_NUM >= 100000)
+		NamedLWLockTranche *namedLockTranche =
+			&distributedTransactionShmemData->namedLockTranche;
+#else
+		LWLockTranche *lockTranche = &distributedTransactionShmemData->lockTranche;
+#endif
+
+		/* start by zeroing out all the memory */
+		memset(distributedTransactionShmemData, 0,
+			   DistributedTransactionManagementShmemSize());
+
+#if (PG_VERSION_NUM >= 100000)
+		namedLockTranche->trancheId = LWLockNewTrancheId();
+
+		LWLockRegisterTranche(namedLockTranche->trancheId, trancheName);
+		LWLockInitialize(&distributedTransactionShmemData->lock,
+						 namedLockTranche->trancheId);
+#else
+		distributedTransactionShmemData->trancheId = LWLockNewTrancheId();
+
+		/* we only need a single lock */
+		lockTranche->array_base = &distributedTransactionShmemData->lock;
+		lockTranche->array_stride = sizeof(LWLock);
+		lockTranche->name = trancheName;
+
+		LWLockRegisterTranche(distributedTransactionShmemData->trancheId, lockTranche);
+		LWLockInitialize(&distributedTransactionShmemData->lock,
+						 distributedTransactionShmemData->trancheId);
+#endif
+
+		/* start the distributed transaction ids from 1 */
+		pg_atomic_init_u64(&distributedTransactionShmemData->nextTransactionId, 1);
+	}
+
+	LWLockRelease(AddinShmemInitLock);
+
+	if (prev_shmem_startup_hook != NULL)
+	{
+		prev_shmem_startup_hook();
+	}
+}
+
+
+/*
+ * DistributedTransactionShmemSize returns the size that should be allocated
+ * on the shared memory for distributed transaction management.
+ */
+static size_t
+DistributedTransactionManagementShmemSize(void)
+{
+	Size size = 0;
+
+	size = add_size(size, sizeof(DistributedTransactionShmemData));
+	size = add_size(size, mul_size(sizeof(DistributedTransactionBackendData),
+								   MaxBackends));
+
+	return size;
+}
+
+
+/*
+ *  InitializeDistributedTransactionManagementBackend is called per backend
+ *  and does the required initilizations.
+ */
+void
+InitializeDistributedTransactionManagementBackend(void)
+{
+	LWLockAcquire(&distributedTransactionShmemData->lock, LW_EXCLUSIVE);
+
+	MyDistributedTransactionBackend =
+		&distributedTransactionShmemData->sessions[MyProc->pgprocno];
+
+	Assert(MyDistributedTransactionBackend);
+
+	MyDistributedTransactionBackend->databaseId = 0;
+	MyDistributedTransactionBackend->transactionId.initiatorNodeIdentifier = 0;
+	MyDistributedTransactionBackend->transactionId.transactionId = 0;
+	MyDistributedTransactionBackend->transactionId.timestamp = 0;
+
+	/* we'll use mutex while accessing the backend shmem */
+	SpinLockInit(&MyDistributedTransactionBackend->mutex);
```

Comment:
This isn't a good idea - this should be initialized *once* in `DistributedTransactionManagementShmemInit()`. Consider what otherwise happens if somebody is currently looking at this entry - you're destroying the spinlock's internal state.   You also should hold the spinlock here while resetting the data.